### PR TITLE
Ported some build fixes from dev branch

### LINKED ISSUE #2124: Master Build warnings with IL2CPP in 2017+
## Overview
The `MetroHandleRef.cs` class conflicts with the imported type 'HandleRef" in mscorlib when using the IL2CPP scripting backend.

## Unity Editor Versions
- 2017.4.3f1
- 2018.1.0f2

## Mixed Reality Toolkit Release Version
84493c975fbf0fe0f3541303ce1758ca11dfe44a

### PR DESCRIPTION
**Pull request by StephenHodgson**
Ported some build fixes and improvements from the dev branch.

- Fixed an issue where users were unable to build without either specifying a wsa certificate, or building through the normal Unity Editor Build window because the certificate had not been generated yet. (Unity Build will generate one for you if one doesn't exist.)
- Build Compatibility with 2018.1
- Detailed log reporting of msbuild output.
- Make quick options HoloLens target make sure to target HoloLens only. #1959
- Removed some compile and build warnings.
- Fixes #2124 and other conflicts with IL2CPP
- Addressed some licensing issues from https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/1904#discussion_r186269929

### CODE REVIEW

File: Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployTools.cs [hunk 1]
```diff
@@ -85,12 +85,22 @@ public static bool BuildSLN(string buildDirectory, bool showDialog = true)
                 WSAUwpSdk = EditorUserBuildSettings.wsaUWPSDK,
 
                 // Configure a post build action that will compile the generated solution
+                // Configure a post build action that will compile the generated solution
```
Comment (davidkline-ms) [inline Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployTools.cs]: This line looks like a duplicate... Is it just Gits's diff?

Comment (StephenHodgson) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployTools.cs]: I'll double check

Comment (StephenHodgson) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployTools.cs]: nope it was a dup.

File: Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployTools.cs [hunk 2]
```diff
@@ -325,15 +363,15 @@ public static bool BuildAppxFromSLN(string productName, string msBuildVersion, b
             return true;
         }
 
-        private static void IncrementPackageVersion()
+        private static bool SetPackageVersion(bool increment)
```
Comment (davidkline-ms) [inline Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployTools.cs]: The new name feels wrong... We cannot actually set a version number, only increment or not...

Comment (davidkline-ms) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployTools.cs]: Maybe call it UpdatePackageVersion?

File: Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs [hunk 3]
```diff
@@ -453,6 +444,15 @@ private void AppxBuildGUI()
             }
 
             var curScriptingBackend = PlayerSettings.GetScriptingBackend(BuildTargetGroup.WSA);
+
+            if (curScriptingBackend == ScriptingImplementation.WinRTDotNET)
+            {
+                EditorGUILayout.HelpBox(".NET Scripting backend is depreciated, please use IL2CPP.", MessageType.Warning);
```
Comment (davidkline-ms) [inline Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs]: When did this recommendation change? Also, is it specific to newer versions of Unity (better IL2CPP support)?

File: Assets/HoloToolkit/BuildAndDeploy/Editor/BuildSLNUtilities.cs [hunk 4]
```diff
@@ -126,32 +136,46 @@ public static void PerformBuild(BuildInfo buildInfo)
...
-                    Directory.CreateDirectory(buildInfo.OutputDirectory);
-                }
+                Directory.CreateDirectory(buildInfo.OutputDirectory);
+            }
+
+            OnPreProcessBuild(buildInfo);
 
-                OnPreProcessBuild(buildInfo);
-                buildError = BuildPipeline.BuildPlayer(
+            EditorUtility.DisplayProgressBar("Build Pipeline", "Herding all the cats...", 0.25f);
```
Comment (davidkline-ms) [inline Assets/HoloToolkit/BuildAndDeploy/Editor/BuildSLNUtilities.cs]: Do we really want "Herding all the cats"? One of our release criteria is globalization and I worry that this may be misunderstood / found offensive.

Comment (StephenHodgson) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildSLNUtilities.cs]: Ahh, sorry about that. It wasn't meant to stay nor be offensive in any way. It was just a way of saying we're gathering all the data.

I'll be sure to update this.

Comment (davidkline-ms) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildSLNUtilities.cs]: No stress. Just me being paranoid :)

File: Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs (hunk 3 above)
Comment (StephenHodgson) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs]: Unity officially made .NET support depreciated: https://blogs.unity3d.com/2018/03/28/updated-scripting-runtime-in-unity-2018-1-what-does-the-future-hold/

File: Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployTools.cs (hunk 2 above)
Comment (StephenHodgson) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployTools.cs]: We're always setting it no matter what the package version is.

Comment (davidkline-ms) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployTools.cs]: Makes sense to me now :)

File: Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs (hunk 3 above)
Comment (davidkline-ms) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs]: For now, can we wrap that in a 2018_1_OR_NEWER? Just to be safe

Comment (StephenHodgson) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs]: Not really sure I agree with the 2018, maybe 2017.

In either case it's completely optional and devs don't have to switch it.

Comment (davidkline-ms) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs]: Maybe just reword it to be "has been deprecated in Unity 2018, please consider using IL2CPP"?

Comment (StephenHodgson) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs]: I'd like devs to submit tickets to Unity against IL2CPP in any version so that we get consistent across any editor version.

Comment (davidkline-ms) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs]: Fair enough...

Comment (keveleigh) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs]: Is "depreciated" here correct? I always thought it was "deprecated."

Comment (keveleigh) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs]: Also, I'm not sure I read in there anywhere that .NET is not longer supported as a backend. All I read is that the .NET 3.5 Scripting Runtime Version is being removed in favor of .NET 4.x and .NET Standard 2.0. Am I missing something? There's no mention of IL2CPP or .NET as backends at all.

Comment (StephenHodgson) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs]: https://docs.unity3d.com/2018.2/Documentation/Manual/windowsstore-dotnet.html

Comment (davidkline-ms) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs]: We may want to hold off on pushing toward IL2CPP until vNext then. Do we really want to change the recommendation in master?

Comment (StephenHodgson) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs]: Yup, just trying to be proactive.

Comment (StephenHodgson) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs]: I strongly disagree.  We def want people to get used to the idea early and I'd like devs to submit tickets to Unity against IL2CPP in any version so that we get consistent across all editor versions.

Comment (StephenHodgson) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs]: Besides it's not like we're _forcing_ them to use IL2CPP. It's just a heads up.

Comment (davidkline-ms) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs]: Then we should say "please consider using" :)

Comment (davidkline-ms) [reply Assets/HoloToolkit/BuildAndDeploy/Editor/BuildDeployWindow.cs]: And let's not forget Kurtis noting the spelling error for "deprecated" :)

File: Assets/HoloToolkit/UX/Scripts/Dialog/DialogShell.cs [hunk 5]
```diff
@@ -28,9 +28,6 @@ public class DialogShell : Dialog
         [SerializeField]
         private TextMesh messageText;
 
-        [SerializeField]
-        private GameObject[] oneButtonSet;
```
Comment (keveleigh) [inline Assets/HoloToolkit/UX/Scripts/Dialog/DialogShell.cs]: @cre8ivepark or @johnppella, is this supposed to be used? Looks like it's not. Are there planned but unimplemented features here?

File: Assets/HoloToolkit/UX/Scripts/Progress/ProgressIndicator.cs [hunk 6]
```diff
@@ -29,8 +29,8 @@ public bool IsLoading
         private IndicatorStyleEnum defaultIndicatorStyle = IndicatorStyleEnum.AnimatedOrbs;
         [SerializeField]
         private ProgressStyleEnum defaultProgressStyle = ProgressStyleEnum.Percentage;
-        [SerializeField]
-        private ProgressMessageStyleEnum defaultMessageStyle = ProgressMessageStyleEnum.Visible;
+        //[SerializeField]
+        //private ProgressMessageStyleEnum defaultMessageStyle = ProgressMessageStyleEnum.Visible;
```
Comment (keveleigh) [inline Assets/HoloToolkit/UX/Scripts/Progress/ProgressIndicator.cs]: @cre8ivepark or @johnppella, is this supposed to be used? Looks like it's not. Are there planned but unimplemented features here?

Comment (johnppella) [reply Assets/HoloToolkit/UX/Scripts/Progress/ProgressIndicator.cs]: oneButtonSet is required when the dialog is going to have only one button.
The oneButtonSet and twoButtonSet variables do not need to be flagged with SerializeField
since they are currently not set in the editor in the example ( this is a leftover from the original implementation from which this example was ported).. They can also be defined in the GenerateButtons() function since they are only used within that func.

File: Assets/HoloToolkit/Utilities/Scripts/ApplicationViewManager.cs [hunk 7]
```diff
@@ -49,7 +49,8 @@ private void Start()
         {
             var viewId = ApplicationView.GetForCurrentView().Id;
             var view = CoreApplication.GetCurrentView();
-            if (CallbackDictionary.TryRemove(viewId, out var cb))
```
Comment (keveleigh) [inline Assets/HoloToolkit/Utilities/Scripts/ApplicationViewManager.cs]: The commit says these are coming from the dev branch, but I don't see these changes there. Do they need to be ported over?

Comment (StephenHodgson) [reply Assets/HoloToolkit/Utilities/Scripts/ApplicationViewManager.cs]: They should be there. I didn't get a chance to test out the app view manager just yet though.

Comment (StephenHodgson) [reply Assets/HoloToolkit/Utilities/Scripts/ApplicationViewManager.cs]: Mostly I just removed the c# language features

Comment (StephenHodgson) [reply Assets/HoloToolkit/Utilities/Scripts/ApplicationViewManager.cs]: But I also didn't solve the issue with IL2CPP builds:

```
Assets\MixedRealityToolkit\Utilities\Scripts\ApplicationViewManager.cs(98,36): error CS1545: Property, indexer, or event 'CoreWindow.VisibilityChanged' is not supported by the language; try directly calling accessor methods 'CoreWindow.add_VisibilityChanged(TypedEventHandler<CoreWindow, VisibilityChangedEventArgs>)' or 'CoreWindow.remove_VisibilityChanged(EventRegistrationToken)'
```

Comment (StephenHodgson) [reply Assets/HoloToolkit/Utilities/Scripts/ApplicationViewManager.cs]: Might just roll back this change.

Comment (davidkline-ms) [reply Assets/HoloToolkit/Utilities/Scripts/ApplicationViewManager.cs]: @StephenHodgson, did you decide to roll back this change?

Comment (StephenHodgson) [reply Assets/HoloToolkit/Utilities/Scripts/ApplicationViewManager.cs]: The changes fix the build issues, but I haven't had a chance to test them

Comment (StephenHodgson) [reply Assets/HoloToolkit/Utilities/Scripts/ApplicationViewManager.cs]: I would like to get this into the next release though

Comment (davidkline-ms) [reply Assets/HoloToolkit/Utilities/Scripts/ApplicationViewManager.cs]: next == may?

Comment (StephenHodgson) [reply Assets/HoloToolkit/Utilities/Scripts/ApplicationViewManager.cs]: hopefully

**Comment by StephenHodgson** [comment]
@davidkline-ms, @keveleigh this PR is ready unless @cre8ivepark and @johnppella have any pending CRs

File: Assets/HoloToolkit/Common/Scripts/CrossPlatform/SystemException.cs [hunk 8]
```diff
@@ -1,7 +1,7 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
-#if UNITY_WSA && !UNITY_EDITOR
+#if UNITY_WSA && !UNITY_EDITOR && !ENABLE_IL2CPP
```
Comment (SimonDarksideJ) [inline Assets/HoloToolkit/Common/Scripts/CrossPlatform/SystemException.cs]: Not having the same #IF as others?  No checking for NetStandard2?

File: Assets/HoloToolkit/Common/Scripts/CrossPlatform/MetroHandleRef.cs [hunk 9]
```diff
@@ -1,7 +1,7 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
-#if UNITY_WSA && !UNITY_EDITOR
+#if UNITY_WSA && !UNITY_EDITOR && !ENABLE_IL2CPP
```
Comment (SimonDarksideJ) [inline Assets/HoloToolkit/Common/Scripts/CrossPlatform/MetroHandleRef.cs]: Not having the same #IF as others?  No checking for NetStandard2?

File: Assets/HoloToolkit/Common/Scripts/CrossPlatform/SystemException.cs (hunk 8 above)
Comment (StephenHodgson) [reply Assets/HoloToolkit/Common/Scripts/CrossPlatform/SystemException.cs]: This is only an issue for il2cpp not bc of the API

File: Assets/HoloToolkit/Common/Scripts/CrossPlatform/MetroHandleRef.cs (hunk 9 above)
Comment (StephenHodgson) [reply Assets/HoloToolkit/Common/Scripts/CrossPlatform/MetroHandleRef.cs]: Same as https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/2127#discussion_r190957707
